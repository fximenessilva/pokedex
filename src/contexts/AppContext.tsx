import React, {
  createContext,
  useReducer,
  useContext,
  Dispatch,
  useEffect,
} from 'react';

import { NAMESPACES } from '../utlils/constants';
import { getter, setter } from '../utlils/localStorageHelpers';

type State = {
  isDarkMode: boolean;
};

type Action = { type: 'SET_DARK_MODE'; payload: boolean };

export const THEME_SELECTED = getter(NAMESPACES.theme);

const defaultInitialState: State = {
  isDarkMode: THEME_SELECTED || false,
};

// Define a prop type for the AppProvider component
type AppProviderProps = {
  children: React.ReactNode;
  initialState?: State; // Add initialState prop here
};

const AppContext = createContext<
  | {
      state: State;
      dispatch: Dispatch<Action>;
      toggleDarkMode: (payload: any) => void;
    }
  | undefined
>(undefined);

const appReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case 'SET_DARK_MODE':
      return { ...state, isDarkMode: action.payload };
    default:
      return state;
  }
};

export const AppProvider: React.FC<AppProviderProps> = ({
  children,
  initialState = defaultInitialState,
}) => {
  const [state, dispatch] = useReducer(appReducer, initialState);

  useEffect(() => {
    if (!THEME_SELECTED) {
      setter(NAMESPACES.theme, state.isDarkMode);
    }
  }, []);

  const toggleDarkMode = (payload: boolean) => {
    dispatch({ type: 'SET_DARK_MODE', payload });
    setter(NAMESPACES.theme, payload);
  };

  return (
    <AppContext.Provider value={{ state, dispatch, toggleDarkMode }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  const context = useContext(AppContext);
  if (context === undefined) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
};
