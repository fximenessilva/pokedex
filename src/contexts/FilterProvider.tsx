import { createContext, useContext, useReducer, ChangeEvent } from 'react';

interface FilterState {
  searchTerm: string;
}

interface UseFilterState {
  filterState: FilterState;
  setSearchTerm: (event: ChangeEvent<HTMLInputElement>) => void;
}

export type FilterAction =
  | { type: 'SET_SEARCHTERM'; payload: string }
  | { type: 'SET_FILTERS_TO_INITIAL' };

export const filterReducer = (
  state: FilterState,
  action: FilterAction
): FilterState => {
  switch (action.type) {
    case 'SET_SEARCHTERM':
      return {
        ...state,
        searchTerm: action.payload,
      };

    case 'SET_FILTERS_TO_INITIAL':
      return {
        ...filterInitialState,
      };
    default:
      return state;
  }
};

export const filterInitialState = {
  searchTerm: '',
};

const FilterContext = createContext<UseFilterState | undefined>(undefined);

export function FilterProvider({ children }: { children: React.ReactNode }) {
  const [filterState, dispatch] = useReducer(filterReducer, filterInitialState);

  const setSearchTerm = (event: ChangeEvent<HTMLInputElement>) => {
    const inputValue = event.target.value;
    dispatch({ type: 'SET_SEARCHTERM', payload: inputValue });
  };

  const contextValue = { filterState, setSearchTerm };

  return (
    <FilterContext.Provider value={contextValue}>
      {children}
    </FilterContext.Provider>
  );
}

export function useFilterState() {
  const context = useContext(FilterContext);
  if (context === undefined) {
    throw new Error('useFilterState must be used within a FilterProvider');
  }
  return context;
}
