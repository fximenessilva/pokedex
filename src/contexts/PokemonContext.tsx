import React, { createContext, useContext, useEffect, useReducer } from 'react';

import { fetchPokemons, fetchPokemonDetail } from '../services/api';
import { Pokemons, Pokemon } from '../utlils/types';

type Action =
  | { type: 'SET_POKEMONS'; payload: Pokemons[] }
  | { type: 'SET_POKEMON'; payload: Pokemon }
  | { type: 'SET_ERROR'; payload: any };

interface PokemonContextProps {
  pokemons: Pokemons[] | [];
  pokemon: Pokemon | null;
  loading: boolean;
  error: string | null;
  dispatch: React.Dispatch<Action>;
  getPokemons: () => Promise<void>;
  getPokemon: (name: string) => Promise<void>;
}

const PokemonContext = createContext<PokemonContextProps | undefined>(
  undefined
);

const pokemonReducer = (
  state: PokemonContextProps,
  action: Action
): PokemonContextProps => {
  switch (action.type) {
    case 'SET_POKEMONS':
      return {
        ...state,
        pokemons: action.payload,
        loading: false,
      };
    case 'SET_POKEMON':
      return {
        ...state,
        pokemon: action.payload,
        loading: false,
      };
    case 'SET_ERROR':
      return {
        ...state,
        error: action.payload,
        loading: false,
      };

    default:
      return state;
  }
};

export const PokemonProvider: React.FC = ({ children }) => {
  const initialState: PokemonContextProps = {
    pokemons: [],
    pokemon: null,
    loading: true,
    error: null,
    dispatch: () => {
      // Placeholder comment
    },
    getPokemons: async () => {
      // Placeholder comment
    },
    getPokemon: async (name) => {
      // Placeholder comment
    },
  };

  const [state, dispatch] = useReducer(pokemonReducer, initialState);

  const getPokemons = async () => {
    try {
      const { status, data } = await fetchPokemons();
      if (status !== 200) {
        dispatch({
          type: 'SET_ERROR',
          payload: 'Error fetching pokemons list!',
        });
        return;
      }
      dispatch({ type: 'SET_POKEMONS', payload: data.results });
    } catch (error) {
      dispatch({ type: 'SET_ERROR', payload: error });
    }
  };

  const getPokemon = async (name: string) => {
    try {
      const { status, data } = await fetchPokemonDetail(name);
      if (status !== 200) {
        dispatch({ type: 'SET_ERROR', payload: 'Error fetching pokemon' });
        return;
      }
      dispatch({ type: 'SET_POKEMON', payload: data });
    } catch (error) {
      dispatch({ type: 'SET_ERROR', payload: error });
    }
  };

  return (
    <PokemonContext.Provider
      value={{ ...state, dispatch, getPokemons, getPokemon }}
    >
      {children}
    </PokemonContext.Provider>
  );
};

export const usePokemonContext = () => {
  const context = useContext(PokemonContext);
  if (!context) {
    throw new Error('usePokemonContext must be used within a PokemonProvider');
  }
  return context;
};
