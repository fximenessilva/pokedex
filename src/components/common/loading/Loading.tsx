import React from 'react';
import styled from 'styled-components';

import { useAppContext } from '../../../contexts/AppContext';

type LoaderWrapperProps = {
  $isDarkMode: boolean;
};

const LoaderWrapper = styled.span<LoaderWrapperProps>`
  width: 48px;
  height: 48px;
  display: inline-block;
  position: relative;

  &::after,
  &::before {
    content: '';
    box-sizing: border-box;
    width: 48px;
    height: 48px;
    border-radius: 50%;
    border: 2px solid ${(props) => (props.$isDarkMode ? '#fff' : '#000')};
    position: absolute;
    left: 0;
    top: 0;
    animation: animloader 2s linear infinite;
  }

  &::after {
    animation-delay: 1s;
  }

  @keyframes animloader {
    0% {
      transform: scale(0);
      opacity: 1;
    }
    100% {
      transform: scale(1);
      opacity: 0;
    }
  }
`;

const Loading = () => {
  const {
    state: { isDarkMode },
  } = useAppContext();

  return <LoaderWrapper $isDarkMode={isDarkMode} />;
};

export default Loading;
