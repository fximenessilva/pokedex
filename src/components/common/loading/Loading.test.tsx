import React from 'react';
import { render } from '@testing-library/react';
import { AppProvider, useAppContext } from '../../../contexts/AppContext';
import Loading from './Loading';

jest.mock('../../../contexts/AppContext', () => {
  const actualModule = jest.requireActual('../../../contexts/AppContext');
  return {
    ...actualModule,
    useAppContext: jest.fn(),
  };
});

describe('Loading', () => {
  beforeEach(() => {
    (useAppContext as jest.Mock).mockReturnValue({
      state: { isDarkMode: true },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with dark mode', () => {
    const { container } = render(
      <AppProvider>
        <Loading />
      </AppProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
    // Add more assertions as needed
  });

  it('renders correctly with light mode', () => {
    (useAppContext as jest.Mock).mockReturnValue({
      state: { isDarkMode: false },
    });

    const { container } = render(
      <AppProvider>
        <Loading />
      </AppProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
