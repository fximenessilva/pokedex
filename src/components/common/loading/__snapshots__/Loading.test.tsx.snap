// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading renders correctly with dark mode 1`] = `
<span
  class="sc-beyTiQ bvceMp"
/>
`;

exports[`Loading renders correctly with light mode 1`] = `
<span
  class="sc-beyTiQ ihEZjn"
/>
`;
